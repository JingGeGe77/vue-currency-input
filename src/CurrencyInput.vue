<template>
  <input
    v-currency="{locale, currency, distractionFree, min, max}"
    :value="formattedValue"
    @change="handleChange">
</template>

<script>
import currencyDirective from './curencyDirective'
import defaultOptions from './defaultOptions'

export default {
  name: 'CurrencyInput',
  directives: {
    currency: currencyDirective
  },
  props: {
    value: {
      type: Number,
      default: null
    },
    locale: {
      type: String,
      default: defaultOptions.locale
    },
    currency: {
      type: String,
      required: true
    },
    distractionFree: {
      type: Boolean,
      default: defaultOptions.distractionFree
    },
    min: {
      type: Number,
      default: defaultOptions.min
    },
    max: {
      type: Number,
      default: defaultOptions.max
    }
  },
  data () {
    return {
      formattedValue: this.value
    }
  },
  watch: {
    value (value) {
      if (!this.$el.$ci.focus) {
        this.$el.dispatchEvent(new CustomEvent('input', { detail: { value } }))
      }
    }
  },
  methods: {
    handleChange ({ detail }) {
      if (detail) {
        this.$emit('input', detail.numberValue)
        this.formattedValue = detail.formattedValue
      }
    }
  }
}
</script>
