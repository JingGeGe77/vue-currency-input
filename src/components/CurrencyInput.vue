<template>
  <input
    :value="formattedValue"
    v-on="listeners">
</template>

<script>
import { createTextMaskInputElement } from 'text-mask-core'
import createNumberMask from '../utils/createNumberMask'
import { parse } from '../utils/formatHelper'

export default {
  name: 'CurrencyInput',
  props: {
    value: {
      type: Number,
      default: null
    },
    locale: {
      type: String,
      default: undefined
    },
    currency: {
      type: String,
      required: true
    },
    distractionFree: {
      type: Boolean,
      default: true
    },
    allowNegative: {
      type: Boolean,
      default: true
    }
  },
  data () {
    return {
      focus: false,
      formattedValue: '',
      numberValue: this.value
    }
  },
  computed: {
    listeners () {
      return {
        ...this.$listeners,
        input: (e) => this.handleInput(e),
        focus: (e) => this.handleFocus(e),
        blur: () => this.handleBlur()
      }
    },
    config () {
      const formattedNumber = this.currencyFormat.format(1234)
      const allowDecimal = formattedNumber.indexOf('0') !== -1
      return {
        prefix: this.focus && this.distractionFree ? '' : formattedNumber.substring(0, formattedNumber.indexOf('1')),
        suffix: this.focus && this.distractionFree ? '' : formattedNumber.substring(formattedNumber.lastIndexOf(allowDecimal ? '0' : '4') + 1),
        decimalSymbol: allowDecimal ? formattedNumber.substr(formattedNumber.indexOf('4') + 1, 1) : null,
        thousandsSeparatorSymbol: formattedNumber.substr(formattedNumber.indexOf('1') + 1, 1),
        allowNegative: this.allowNegative,
        allowDecimal
      }
    },
    textMaskInputElement () {
      return createTextMaskInputElement({
        inputElement: this.$el,
        guide: false,
        mask: createNumberMask(this.config)
      })
    },
    currencyFormat () {
      return new Intl.NumberFormat(this.locale, {
        style: 'currency',
        currency: this.currency
      })
    },
    fixedFractionNumberFormat () {
      return new Intl.NumberFormat(this.locale, { minimumFractionDigits: this.config.allowDecimal ? 2 : 0 })
    }
  },
  watch: {
    value (value) {
      if (!this.focus) {
        this.updateValue(this.fixedFractionNumberFormat.format(value))
      }
    },
    allowNegative () {
      this.updateValue(this.formattedValue)
    },
    locale: 'applyFixedFractionFormat',
    currency: 'applyFixedFractionFormat'
  },
  mounted () {
    this.applyFixedFractionFormat()
  },
  methods: {
    handleInput (e) {
      this.updateValue(e.target.value)
    },
    handleBlur () {
      this.focus = false
      this.applyFixedFractionFormat()
    },
    handleFocus (e) {
      this.$nextTick(() => {
        const caretPosition = e.target.selectionStart - this.config.prefix.length
        this.focus = true
        if (this.numberValue !== null && this.distractionFree) {
          this.format(new Intl.NumberFormat(this.locale).format(this.numberValue))
          this.setCaretPosition(Math.max(0, caretPosition))
        }
      })
    },
    updateValue (value) {
      this.format(value)
      this.numberValue = parse(this.formattedValue, this.config)
      this.$emit('input', this.numberValue)
    },
    format (value) {
      this.textMaskInputElement.update(value)
      this.formattedValue = this.$el.value
    },
    applyFixedFractionFormat () {
      if (this.numberValue === null) {
        this.formattedValue = ''
      } else {
        this.format(this.fixedFractionNumberFormat.format(this.numberValue.toFixed(2)))
      }
    },
    setCaretPosition (caretPosition) {
      this.$el.setSelectionRange(caretPosition, caretPosition)
    }
  }
}
</script>
