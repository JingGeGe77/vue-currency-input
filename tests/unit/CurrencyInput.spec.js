import { shallowMount } from '@vue/test-utils'
import CurrencyInput from '../../src/components/CurrencyInput'

describe('CurrencyInput', () => {
  let wrapper, propsData
  beforeEach(() => {
    propsData = {
      currency: 'EUR',
      locale: 'en'
    }
    wrapper = mountComponent(propsData)
  })

  it('is a Vue component', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })

  describe('props', () => {
    describe('value', () => {
      it('is of type Number', () => {
        expect(wrapper.vm.$options.props.value.type).toBe(Number)
      })

      it('defaults to null', () => {
        expect(wrapper.vm.$options.props.value.default).toBeNull()
      })
    })

    describe('currency', () => {
      it('is of type String', () => {
        expect(wrapper.vm.$options.props.currency.type).toBe(String)
      })

      it('is required', () => {
        expect(wrapper.vm.$options.props.currency.required).toBe(true)
      })
    })

    describe('locale', () => {
      it('is of type String', () => {
        expect(wrapper.vm.$options.props.locale.type).toBe(String)
      })

      it('defaults to undefined', () => {
        expect(wrapper.vm.$options.props.locale.required).toBe(undefined)
      })
    })

    describe('distractionFree', () => {
      it('is of type Boolean', () => {
        expect(wrapper.vm.$options.props.distractionFree.type).toBe(Boolean)
      })

      it('defaults to true', () => {
        expect(wrapper.vm.$options.props.distractionFree.default).toBe(true)
      })
    })

    describe('allowNegative', () => {
      it('is of type Boolean', () => {
        expect(wrapper.vm.$options.props.allowNegative.type).toBe(Boolean)
      })

      it('defaults to true', () => {
        expect(wrapper.vm.$options.props.allowNegative.default).toBe(true)
      })
    })
  })

  describe('when the component is mounted', () => {
    it('applies the fixed fraction format to the initial value', () => {
      const applyFixedFractionFormat = jest.fn()
      shallowMount(CurrencyInput, {
        propsData,
        methods: { applyFixedFractionFormat }
      })
      expect(applyFixedFractionFormat).toHaveBeenCalled()
    })
  })

  describe('currency format config', () => {
    let format
    beforeEach(() => {
      format = jest.fn()
      wrapper = shallowMount(CurrencyInput, {
        propsData,
        computed: {
          currencyFormat: jest.fn(() => ({ format }))
        }
      })
    })

    describe('when the format has a prefix', () => {
      it('builds the config correctly', () => {
        format.mockReturnValue('€ 1,234.00')

        expect(wrapper.vm.config).toMatchSnapshot()
      })
    })

    describe('when the format has a suffix', () => {
      it('builds the config correctly', () => {
        format.mockReturnValue('1.234,00 €')

        expect(wrapper.vm.config).toMatchSnapshot()
      })
    })
  })

  describe('when the input is changed', () => {
    it('formats the value correctly', () => {
      wrapper.element.value = '1234'

      wrapper.trigger('input')

      expect(wrapper.vm.formattedValue).toBe('€1,234')
    })

    it('emits the raw number value', () => {
      wrapper.element.value = '€1,234.5'

      wrapper.trigger('input')

      expect(wrapper.emitted('input')[0][0]).toBe(1234.5)
    })

    describe('when the input is cleared', () => {
      it('emits a null value', () => {
        wrapper.element.value = ''

        wrapper.trigger('input')

        expect(wrapper.emitted('input')[0][0]).toBeNull()
      })
    })
  })

  describe('when the input is focused', () => {
    describe('distraction free mode is enabled', () => {
      it('applies the distraction free format', async () => {
        wrapper.setProps({ value: 1234.5, distractionFree: true })

        wrapper.trigger('focus')
        await wrapper.vm.$nextTick()

        expect(wrapper.vm.formattedValue).toBe('1,234.5')
      })

      it('sets the caret position correctly', async () => {
        wrapper.setProps({ value: 1234.5, distractionFree: true })
        jest.spyOn(wrapper.vm, 'setCaretPosition')

        wrapper.element.setSelectionRange(1, 1)
        wrapper.trigger('focus')
        await wrapper.vm.$nextTick()

        expect(wrapper.vm.setCaretPosition).toHaveBeenCalledWith(0)
      })
    })

    describe('distraction free mode is disabled', () => {
      it('leaves the current value untouched', async () => {
        wrapper.setProps({ value: 1234.5, distractionFree: false })

        wrapper.trigger('focus')
        await wrapper.vm.$nextTick()

        expect(wrapper.vm.formattedValue).toBe('€1,234.50')
      })

      it('leaves the caret position untouched', async () => {
        wrapper.setProps({ value: 1234.5, distractionFree: false })
        jest.spyOn(wrapper.vm, 'setCaretPosition')

        wrapper.element.setSelectionRange(3, 3)
        wrapper.trigger('focus')
        await wrapper.vm.$nextTick()

        expect(wrapper.element.selectionStart).toBe(3)
      })
    })
  })

  describe('when the input is blurred', () => {
    it('applies the fixed fraction format', () => {
      wrapper.setProps({ value: 1234.5 })

      wrapper.trigger('blur')

      expect(wrapper.vm.formattedValue).toBe('€1,234.50')
    })
  })
})

const mountComponent = (propsData) => {
  return shallowMount(CurrencyInput, { propsData })
}
