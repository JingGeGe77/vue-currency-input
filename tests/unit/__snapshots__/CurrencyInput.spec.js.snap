// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput currency format config when the format has a prefix builds the config correctly 1`] = `
Object {
  "allowDecimal": true,
  "allowNegative": true,
  "decimalSymbol": ".",
  "prefix": "€ ",
  "suffix": "",
  "thousandsSeparatorSymbol": ",",
}
`;

exports[`CurrencyInput currency format config when the format has a suffix builds the config correctly 1`] = `
Object {
  "allowDecimal": true,
  "allowNegative": true,
  "decimalSymbol": ",",
  "prefix": "",
  "suffix": " €",
  "thousandsSeparatorSymbol": ".",
}
`;
